File: leetcode/symmetric_tree.go
// https://leetcode.com/problems/symmetric-tree/
// 101. Symmetric Tree
package main

import "fmt"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func MakeTreeNode(val int, left, right *TreeNode) *TreeNode {
	return &TreeNode{
		Val:   val,
		Left:  left,
		Right: right,
	}
}

func isSymmetric(root *TreeNode) bool {
	res := true
	res = compare(root, root)
	return res
}
func compare(sRoot1 *TreeNode, sRoot2 *TreeNode) bool {
	if sRoot1 == nil && sRoot2 == nil {
		return true
	} else if sRoot1 == nil || sRoot2 == nil {
		return false
	} else if sRoot1.Val != sRoot2.Val {
		return false
	}
	return compare(sRoot1.Left, sRoot2.Right) && compare(sRoot1.Right, sRoot2.Left)
}

func main() {
	input := []int{1, 2, 2, 3, 4, 4, 3}
	tree := makeTree(input)
	res := isSymmetric(tree)
	fmt.Println(res)
}

func makeTree(input []int) *TreeNode {
	return getTree(0, input)
}
func getTree(subRootIndex int, input []int) *TreeNode {
	if subRootIndex >= len(input) || input[subRootIndex] == -1 {
		return &TreeNode{}
	}
	return MakeTreeNode(input[subRootIndex], getTree((subRootIndex*2)+1, input), getTree((subRootIndex+1)*2, input))
}

File: leetcode/symmetric_tree.py
# Definition for a binary tree node.
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:

        def recurse(left, right) -> bool:
            if left is None and right is None:
                return True
            if left is None or right is None:
                return False
            if left.val != right.val:
                return False
            return recurse(left.left, right.right) and recurse(left.right, right.left)

        return recurse(root.left, root.right)


def getTree(nodes: List[int]) -> TreeNode:
    def recurse(index):
        if index >= len(nodes):
            return None
        if nodes[index] is None:
            return None
        subroot = TreeNode(nodes[index])
        left = recurse((index * 2) + 1)
        right = recurse((index * 2) + 2)
        subroot.left = left
        subroot.right = right
        return subroot

    return recurse(0)


if __name__ == '__main__':
    list = [1, 2, 2, 3, 3, 3, 3]
    tree = getTree(list)
    res = Solution().isSymmetric(tree)
    print(res)
